File: _REPO/APRESS/distributed-.net-programming-in-vb-.net/Chapter04/DisposingSponsor/MathLibrary/Customer.vb
Imports System.Runtime.Remoting.Lifetime 'ILease
Imports System.Threading

Public Class Customer
   Inherits MarshalByRefObject
   Implements IDisposable

   Private mName As String

   Public Sub New()
      Console.WriteLine("Customer.ctor()")
   End Sub

   Public Sub New(name As String)
      Console.WriteLine("Customer.ctor({0}) on thread {1}", name, _
         Thread.CurrentThread.GetHashCode())
      mName = name
   End Sub

   Public Function SayHello() As String
      Console.WriteLine("Customer.SayHello() on thread {0}", _
         Thread.CurrentThread.GetHashCode())
      Return "Hello " & mName
   End Function

   Public Overrides Function InitializeLifetimeService() As Object
      'Call base class version
      Dim leaseInfo As ILease
      leaseInfo = CType(MyBase.InitializeLifetimeService(), ILease)

      'Register a CustomerSponsor object as a sponsor.
      leaseInfo.Register(New CustomerSponsor())
      
      'Register a DisposingSponsor object
      leaseInfo.Register(New DisposingSponsor(Me))
      
      Return leaseInfo
   End Function

   Public Sub Dispose() Implements IDisposable.Dispose
      Console.WriteLine("Customer.Dispose() on thread {0}", _
         Thread.CurrentThread.GetHashCode())
      GC.SuppressFinalize(Me)
   End Sub

   Protected Overrides Sub Finalize()
      Console.WriteLine("Customer.Finalize()")
      MyBase.Finalize()
   End Sub
End Class

Class CustomerSponsor
   Implements ISponsor

   Private mRenewCount As Integer = 0

   Public Function Renewal(leaseInfo As ILease) As TimeSpan _
      Implements ISponsor.Renewal

      'Just a trace message for testing
      Console.WriteLine("CustomerSponsor.Renewal()")

      If mRenewCount < 3 Then
         mRenewCount += 1
         Return TimeSpan.FromSeconds(5)
      Else
         Return TimeSpan.FromSeconds(0)
      End If
   End Function

End Class

File: _REPO/APRESS/distributed-.net-programming-in-csharp/Chapter04/DisposingSponsor/MathLibrary/Customer.cs
using System;
using System.Runtime.Remoting.Lifetime;
using System.Threading;

namespace MathLibrary
{
   public class Customer : MarshalByRefObject, IDisposable
   {
      string mName;

      public Customer()
      {
         Console.WriteLine("Customer.ctor()");
      }

      public Customer(string name)
      {
         Console.WriteLine("Customer.ctor({0}) on thread {1}", name,
            Thread.CurrentThread.GetHashCode());
         mName = name;
      }

      public string SayHello()
      {
         Console.WriteLine("Customer.SayHello() on thread {0}", 
            Thread.CurrentThread.GetHashCode());

         return "Hello " + mName;
      }

      public override object InitializeLifetimeService()
      {  
         ILease leaseInfo = (ILease)base.InitializeLifetimeService();
         
         // Register a CustomerSponsor object as a sponsor.
         leaseInfo.Register(new CustomerSponsor());

         // Register a DisposingSponsor object
         leaseInfo.Register(new DisposingSponsor(this));
         return leaseInfo;
      }

      // Implements IDisposable.Dispose()
      public void Dispose()
      {
         Console.WriteLine("Customer.Dispose() on thread {0}", 
            Thread.CurrentThread.GetHashCode()
         );
         GC.SuppressFinalize(this);
      }

      // The Finalizer
      ~Customer()
      {
         Console.WriteLine("Customer.Finalize()");
      }
   }

   class CustomerSponsor : ISponsor
   {
      private int mRenewCount = 0;
      
      // Implements ISponsor.Renewal
      public TimeSpan Renewal(ILease leaseInfo)
      {
         if (mRenewCount < 3)
         {      
            mRenewCount++;

            // Just a trace message for testing
            Console.WriteLine("CustomerSponsor.Renewal()");
            return TimeSpan.FromSeconds(5);
         } 
         else 
         {
            Console.WriteLine("CustomerSponsor.Renewal() returning 0 time!");
            return TimeSpan.FromSeconds(0);
         }
      }
   }
}
